layer/layer_data.cc

```diff
@@ -72,15 +103,24 @@ LayerData::LayerData(char* log_filename, const char* header) {
   if (log_filename) {
     out_ = fopen(log_filename, "w");
     if (out_ == nullptr) {
-      fprintf(stderr, "Failed to open %s, output will be to STDERR.\n",
-              log_filename);
+      LOG(ERROR) << "Failed to open " << log_filename
+                 << ", output will be to STDERR.\n";
       out_ = stderr;
     }
   } else {
     out_ = stderr;
   }
-  fprintf(out_, "%s\n", header);
-  fflush(out_);
+
+  WriteLnAndFlush(out_, header);
+
+  if (const char* event_log_file = getenv(kEventLogFileEnvVar)) {
+    absl::MutexLock lock(&event_log_lock_);
+    // We should probably use API that allows us to lock the event log file so
+    // that every line is written atomically and flushed. Printing and fflushing
+    // seems to be working fine at the moment, but we can revisit this in the
+    // future.
```

Comment:
I'd remove or reword this comment to simply say all lines should be written in a single fprintf call (the FILE functions are atomic within a process, as in two fprintf's or other calls to the same FILE* will internally take a lock).

You've refactored all the fprintf() calls to use WriteLnAndFlush which ensures all writes are done in single fprintf calls - because of this log_lock_ and event_log_lock_ can be removed entirely (they could before as well AFAICT, but now it's clearer that all writes are full lines)

Reply:
Done. Nice observation.